File: src/main/java/frc/robot/subsystems/superstructure/Superstructure.java
package frc.robot.subsystems.superstructure;

import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
import edu.wpi.first.wpilibj2.command.Command;
import edu.wpi.first.wpilibj2.command.Commands;
import edu.wpi.first.wpilibj2.command.SubsystemBase;
import edu.wpi.first.wpilibj2.command.button.Trigger;
import edu.wpi.first.wpilibj2.command.sysid.SysIdRoutine.Direction;
import frc.robot.Constants.ManipulatorConstants;
import frc.robot.subsystems.superstructure.arm.Arm;
import frc.robot.subsystems.superstructure.manipulator.Manipulator;
import frc.robot.subsystems.swerve.Drivebase;
import java.util.function.BooleanSupplier;
import java.util.function.Supplier;
import org.littletonrobotics.junction.Logger;

/** Coordinates arm and manipulator mechanisms. */
public final class Superstructure extends SubsystemBase {
  private final Drivebase drivebase;
  private final Arm arm;
  private final Manipulator manipulator;

  private SuperstructureState lastRequest = SuperstructureGoals.STOW;

  private final Trigger isL1 = new Trigger(() -> lastRequest == SuperstructureGoals.L1);
  private final Trigger readyToScore = new Trigger(() -> SmartDashboard.getBoolean("shouldScore", false));

  public Superstructure(Drivebase drivebase) {
    this.drivebase = drivebase;
    this.arm = new Arm();
    this.manipulator = new Manipulator();
  }

  public Command getArmSysIdCommand() {
    return Commands.sequence(
        arm.getDynamicSysID(Direction.kForward),
        arm.getDynamicSysID(Direction.kReverse),
        arm.getQuasistaticSysID(Direction.kForward),
        arm.getQuasistaticSysID(Direction.kReverse));
  }

  public Trigger isL1() {
    return isL1;
  }

  public Trigger isHoldingCoral() {
    double threshold = ManipulatorConstants.DETECTION_DISTANCE_THRESHOLD_METERS;
    return new Trigger(() -> manipulator.getInputs().coralDistanceMeters > threshold);
  }

  public Command getSendToStateCommand(Supplier<SuperstructureState> stateSupplier) {
    return Commands.sequence(
        getForceStateCommand(stateSupplier),
        arm.getMoveToAngleCommand(() -> stateSupplier.get().getArmAngle()));
  }

  public Command getForceStateCommand(Supplier<SuperstructureState> stateSupplier) {
    return Commands.runOnce(
        () -> {
          lastRequest = stateSupplier.get();
          drivebase.setLimits(lastRequest.getDriveLimits());
        });
  }

  public Trigger getStateAchievedTrigger(SuperstructureState target) {
    return new Trigger(
        () ->
            Math.abs(
                    arm.getAngle().minus(target.getArmAngle()).getDegrees())
                < 2.0);
  }

  public Command getScoreCommand(BooleanSupplier shouldScore) {
    return Commands.sequence(
        getSendToStateCommand(() -> lastRequest),
        Commands.waitUntil(() -> getStateAchievedTrigger(lastRequest).getAsBoolean()),
        Commands.waitUntil(() -> !shouldScore.getAsBoolean()),
        Commands.waitUntil(shouldScore::getAsBoolean),
        manipulator.getSpinRollersOutCommand(),
        Commands.waitSeconds(1.0),
        manipulator.getStopRollersCommand());
  }

  public Command getProcessorCommand(BooleanSupplier shouldScore) {
    return Commands.sequence(
        Commands.waitUntil(() -> !shouldScore.getAsBoolean()),
        manipulator.getSpinRollersInCommand(),
        Commands.waitUntil(shouldScore::getAsBoolean),
        manipulator.getStopRollersCommand());
  }

  public Command getIntakeAlgaeCommand(BooleanSupplier done) {
    return Commands.sequence(
        getSendToStateCommand(() -> SuperstructureGoals.ALGAE_INTAKE),
        Commands.waitUntil(() -> getStateAchievedTrigger(SuperstructureGoals.ALGAE_INTAKE).getAsBoolean()),
        manipulator.getSpinRollersOutCommand(),
        Commands.waitUntil(() -> !done.getAsBoolean()),
        Commands.waitUntil(done::getAsBoolean),
        manipulator.getStopRollersCommand());
  }

  public Command getIntakeCommand(BooleanSupplier shouldRetract) {
    return Commands.sequence(
        getSendToStateCommand(() -> SuperstructureGoals.GROUND),
        Commands.waitUntil(() -> getStateAchievedTrigger(SuperstructureGoals.GROUND).getAsBoolean()),
        Commands.waitUntil(() -> !shouldRetract.getAsBoolean()),
        manipulator.getSpinRollersInCommand(),
        Commands.waitUntil(shouldRetract::getAsBoolean),
        manipulator.getStopRollersCommand(),
        getSendToStateCommand(() -> SuperstructureGoals.STOW),
        Commands.waitUntil(() -> getStateAchievedTrigger(SuperstructureGoals.STOW).getAsBoolean()));
  }

  public Command getDealgaefyL2Command(BooleanSupplier readyToDealgaefy) {
    return Commands.sequence(
        Commands.waitUntil(() -> !readyToDealgaefy.getAsBoolean()),
        getSendToStateCommand(() -> SuperstructureGoals.PRE_ALGAE_L2),
        Commands.waitUntil(() -> getStateAchievedTrigger(SuperstructureGoals.PRE_ALGAE_L2).getAsBoolean()),
        Commands.deadline(
            Commands.waitUntil(readyToDealgaefy::getAsBoolean),
            manipulator.getSpinRollersOutCommand()),
        manipulator.getStopRollersCommand());
  }

  public Command getDealgaefyL3Command(BooleanSupplier readyToDealgaefy) {
    return Commands.sequence(
        Commands.waitUntil(() -> !readyToDealgaefy.getAsBoolean()),
        getSendToStateCommand(() -> SuperstructureGoals.PRE_ALGAE_L3),
        Commands.waitUntil(() -> getStateAchievedTrigger(SuperstructureGoals.PRE_ALGAE_L3).getAsBoolean()),
        Commands.deadline(
            Commands.waitUntil(readyToDealgaefy::getAsBoolean),
            manipulator.getSpinRollersInCommand()),
        manipulator.getStopRollersCommand());
  }

  @Override
  public void periodic() {
    Logger.recordOutput("superstructure/setpoint", SuperstructureState.STRUCT, lastRequest);
    Logger.recordOutput("superstructure/setpoint/name", lastRequest.getName());
  }

  public Drivebase getDrivebase() {
    return drivebase;
  }

  public Arm getArm() {
    return arm;
  }

  public Manipulator getManipulator() {
    return manipulator;
  }

  public SuperstructureState getLastRequest() {
    return lastRequest;
  }
}

File: src/main/java/frc/robot/subsystems/superstructure/Superstructure.kt
package frc.robot.subsystems.superstructure

import edu.wpi.first.units.Units.*
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard
import edu.wpi.first.wpilibj2.command.Command
import edu.wpi.first.wpilibj2.command.Commands
import edu.wpi.first.wpilibj2.command.Commands.runOnce
import edu.wpi.first.wpilibj2.command.button.Trigger
import edu.wpi.first.wpilibj2.command.sysid.SysIdRoutine
import frc.robot.Robot.drivebase
import frc.robot.subsystems.superstructure.SuperstructureGoals.L1
// import frc.robot.subsystems.superstructure.SuperstructureGoals.L2
// import frc.robot.subsystems.superstructure.SuperstructureGoals.L3
// import frc.robot.subsystems.superstructure.SuperstructureGoals.L4
import frc.robot.subsystems.superstructure.SuperstructureGoals.STOW
// import frc.robot.subsystems.superstructure.elevator.Elevator
import frc.robot.subsystems.superstructure.arm.Arm
import frc.robot.subsystems.superstructure.manipulator.Manipulator
import lib.math.units.degrees
import lib.math.units.radians
import lib.math.units.epsilonEquals
import lib.math.units.inches
import org.littletonrobotics.junction.Logger
import java.util.function.BooleanSupplier
import java.util.function.Supplier

class Superstructure {
    val arm: Arm
    val manipulator: Manipulator

    var lastRequest: SuperstructureState = SuperstructureGoals.STOW

    init {
        arm = Arm()
        manipulator = Manipulator()

        // manipulator.apply {
        //     defaultCommand = this.run {
        //         if (inputs.coralDistance > Meters.of(0.2)) {
        //             io.setRollerVoltage(Volts.of(1.0))
        //         } else {
        //             io.setRollerVoltage(Volts.zero())
        //         }
        //     }.onlyIf { lastRequest == STOW }.handleInterrupt { io.setRollerVoltage(Volts.zero()) }
        // }

        // elevator.apply {
        //     defaultCommand = this.run {
        //         if(lastRequest == STOW && manipulator.inputs.coralDistance < Meters.of(0.2)) {
        //             io.setElevatorHeightTarget(STOW.elevatorHeight + 12.0.inches)
        //         } else {
        //             io.setElevatorHeightTarget(STOW.elevatorHeight)
        //         }
        //     }
        // }
    }


    fun getArmSysIDCommand(): Command {
        return Commands.sequence(
            arm.getDynamicSysID(SysIdRoutine.Direction.kForward),
            arm.getDynamicSysID(SysIdRoutine.Direction.kReverse),
            arm.getQuasistaticSysID(SysIdRoutine.Direction.kForward),
            arm.getQuasistaticSysID(SysIdRoutine.Direction.kReverse)
        )
    }

    // fun getElevatorSysIDCommand(): Command {
    //     return Commands.sequence(
    //         elevator.getDynamicSysID(SysIdRoutine.Direction.kForward),
    //         elevator.getDynamicSysID(SysIdRoutine.Direction.kReverse),
    //         elevator.getQuasistaticSysID(SysIdRoutine.Direction.kForward),
    //         elevator.getQuasistaticSysID(SysIdRoutine.Direction.kReverse)
    //     )
    // }

    // fun getManipulatorSysIDCommand(): Command {
    //     return Commands.sequence(
    //         manipulator.getDynamicTest(SysIdRoutine.Direction.kForward),
    //         manipulator.getDynamicTest(SysIdRoutine.Direction.kReverse),
    //         manipulator.getQuasistaticTest(SysIdRoutine.Direction.kForward),
    //         manipulator.getQuasistaticTest(SysIdRoutine.Direction.kReverse)
    //     )
    // }

    private val readyToScore: Trigger = Trigger { SmartDashboard.getBoolean("shouldScore", false) }

    val isL1: Trigger = Trigger { lastRequest == L1 }
    // val isL2: Trigger = Trigger { lastRequest == L2 }
    // val isL3: Trigger = Trigger { lastRequest == L3 }
    // val isL4: Trigger = Trigger { lastRequest == L4 }

    val isHoldingCoral: Trigger = Trigger { manipulator.inputs.coralDistance > Meters.of(0.2) }

    fun getSendToStateCommand(superstructureState: Supplier<SuperstructureState>): Command {
        return Commands.sequence(
            getForceStateCommand(superstructureState),
            arm.getMoveToAngleCommand { superstructureState.get().armAngle },
        )
    }

    fun getForceStateCommand(stateSupplier: Supplier<SuperstructureState>): Command {
        return runOnce({
            lastRequest = stateSupplier.get()
            drivebase.limits = lastRequest.driveLimits
        })
    }

    fun getStateAchievedTrigger(target: SuperstructureState): Trigger {
        return Trigger {
            Logger.recordOutput("target", target.name)
            val armAtSetpoint = arm.inputs.angle.epsilonEquals(
                target.armAngle,
                5.0.degrees
            )
            return@Trigger armAtSetpoint
        }
    }

    /// when shouldScore is true, coral will be released
    fun getScoreCommand(shouldScore: BooleanSupplier): Command {
        return Commands.sequence(
            getSendToStateCommand { lastRequest },
            Commands.waitUntil { getStateAchievedTrigger(lastRequest).asBoolean },
            Commands.waitUntil { shouldScore.getAsBoolean() },
            manipulator.getSpinRollersOutCommand(),
            Commands.waitSeconds(1.0),
            manipulator.getStopRollersCommand(),
            getSendToStateCommand { SuperstructureGoals.STOW },
            Commands.waitUntil { getStateAchievedTrigger(SuperstructureGoals.STOW).asBoolean },
        )
    }

    // new intake command
    fun getIntakeCommand(): Command {
        return Commands.sequence(
            getSendToStateCommand { SuperstructureGoals.GROUND },
            Commands.waitUntil { getStateAchievedTrigger(SuperstructureGoals.GROUND).asBoolean },
            manipulator.getSpinRollersInCommand(),
            Commands.waitSeconds(3.0),
            manipulator.getStopRollersCommand(),
        )
    }

    // fun getDealgaefyCommand(): Command {
    //     // when run, grab algae and return to ALGAE_STOW with manipulator rollers in brake mode
    //     return Commands.sequence(
    //         getSendToStateCommand { SuperstructureGoals.ALGAE_L2 }.onlyIf{lastRequest == SuperstructureGoals.L2},
    //         getSendToStateCommand { SuperstructureGoals.ALGAE_L3 }.onlyIf{lastRequest == SuperstructureGoals.L3},
    //         Commands.waitUntil { getStateAchievedTrigger(SuperstructureGoals.ALGAE_L2).asBoolean }.onlyIf{lastRequest == SuperstructureGoals.L2},
    //         Commands.waitUntil { getStateAchievedTrigger(SuperstructureGoals.ALGAE_L3).asBoolean }.onlyIf{lastRequest == SuperstructureGoals.L3},
    //         // manipulator.getSpinRollersOutCommand().until(!manipulator.isDetectingGamePiece())), // could be in?
    //         Commands.deadline( // placeholder for sim
    //             Commands.waitSeconds(3.0),
    //             manipulator.getSpinRollersOutCommand(),
    //         ),
    //         manipulator.getStopRollersCommand(),
    //         getSendToStateCommand { SuperstructureGoals.ALGAE_STOW },
    //         Commands.waitUntil { getStateAchievedTrigger(SuperstructureGoals.ALGAE_STOW).asBoolean },
    //     )
    // }

    // todo, will need to release while moving up to throw algae
    // fun getReleaseAlgaeCommand(): Command {
    //     return Commands.sequence(
    //         getSendToStateCommand { lastRequest },
    //         Commands.waitUntil { getStateAchievedTrigger(lastRequest).asBoolean },
    //         manipulator.getSpinRollersInCommand(),
    //         Commands.waitSeconds(10.0),
    //         manipulator.getStopRollersCommand(),
    //         getSendToStateCommand { SuperstructureGoals.STOW },
    //         Commands.waitUntil { getStateAchievedTrigger(SuperstructureGoals.STOW).asBoolean },
    //     )
    // }

    fun periodic() {
        Logger.recordOutput("superstructure/setpoint", SuperstructureState.struct, lastRequest)
        Logger.recordOutput("superstructure/setpoint/name", lastRequest.name)
    }
}